<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-общей-информации" data-toc-modified-id="Изучение-общей-информации-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение общей информации</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Проведём-исследование-данных-каждого-столбца" data-toc-modified-id="Проведём-исследование-данных-каждого-столбца-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Проведём исследование данных каждого столбца</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Категоризация данных</a></span></li><li><span><a href="#Содание-дополнительных-таблиц" data-toc-modified-id="Содание-дополнительных-таблиц-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Содание дополнительных таблиц</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы,-поставленные-в-начале-исследования" data-toc-modified-id="Ответы-на-вопросы,-поставленные-в-начале-исследования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы, поставленные в начале исследования</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Исследование надёжности заёмщиков

Необходимо получить ответы на следующие вопросы:

•	Есть ли зависимость между наличием детей и возвратом кредита в срок?

•	Есть ли зависимость между семейным положением и возвратом кредита в срок?

•	Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

•	Как разные цели кредита влияют на его возврат в срок?

## Изучение общей информации

**Описание данных**

•	children — количество детей в семье

•	days_employed — общий трудовой стаж в днях

•	dob_years — возраст клиента в годах

•	education — уровень образования клиента

•	education_id — идентификатор уровня образования

•	family_status — семейное положение

•	family_status_id — идентификатор семейного положения

•	gender — пол клиента

•	income_type — тип занятости

•	debt — имел ли задолженность по возврату кредитов

•	total_income — ежемесячный доход

•	purpose — цель получения кредита

Импортируем библиотеки, прочитаем файл, получим информацию о самом файле, его размерах, типах данных

In [1]:
import pandas as pd
from nltk.stem import SnowballStemmer
russian_stemmer = SnowballStemmer('russian')

df = pd.read_csv('/datasets/data.csv')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [2]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [3]:
df.shape

(21525, 12)

In [4]:
# получим дополнительные сведения о датафрейме
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Вывод из ознакомительной части: таблица содержит 21525 - строк, 12 столбцов. Большая часть числовых данных имеет тип int64(5), 2 столбца - float64 (в них число строк меньше и составляет  19351). Столбец  children — количество детей в семье - имеет отрицательное значение (-1), а также максимальное значение - 20 (что, скорее всего, является некорректным вводом данных); days_employed — общий трудовой стаж в днях - имеет отрицательные значения,а также максимальное значение - 401755.400475 дней!(вероятно количество указано в часах)

In [5]:
df.describe(include=['int64'])

children     dob_years  education_id  family_status_id  \
count  21525.000000  21525.000000  21525.000000      21525.000000   
mean       0.538908     43.293380      0.817236          0.972544   
std        1.381587     12.574584      0.548138          1.420324   
min       -1.000000      0.000000      0.000000          0.000000   
25%        0.000000     33.000000      1.000000          0.000000   
50%        0.000000     42.000000      1.000000          0.000000   
75%        1.000000     53.000000      1.000000          1.000000   
max       20.000000     75.000000      4.000000          4.000000   

               debt  
count  21525.000000  
mean       0.080883  
std        0.272661  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

## Предобработка данных

### Проведём исследование данных каждого столбца 

2.1.Получим информацию об уникальных значениях столбца children — количество детей в семье

In [6]:
print(df['children'].value_counts()) # вернёт уникальные значения и их количество;
print(df['children'].isna().sum()) # вернёт количество пропусков;
print(len(df['children'])) # покажет длину столбца

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64
0
21525


**Вывод по children**: Количество строк соответствует общему количеству в датафрейме, пропуски отсутствуют, необходимо внести корректировку; доля отрицательного значения(-1) - 0,2%, полагаю, что имеет место "человеческий фактор" - просто заменим -1 на 1; доля значения 20 - 0,3% -"человеческий фактор" (может быть соскользнула рука при вводе, 2 и 0 расположены друг под другом). Заменим 20 на 2.

In [7]:
# чтобы не повредить исходный датафрейм создадим рабочую версию
df_work = df

In [8]:
# произведём замену в столбце children
df_work.loc[df_work['children'] == -1, 'children'] = 1
df_work.loc[df_work['children'] == 20, 'children'] = 2
print(df['children'].value_counts())

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64


In [9]:
# Доля каждой группы:
df_work.groupby(df_work['children']).size() / len(df_work['children'])

children
0    0.657329
1    0.226016
2    0.099001
3    0.015331
4    0.001905
5    0.000418
dtype: float64

После замены данные вполне реалистичные. Наиболее многочисленная группа - лица, не имеющие детей - 66%, далее, с одним ребёнком - 22,6%, с двумя детьми - 10%

2.2.Получим информацию об уникальных значениях столбца days_employed - общий трудовой стаж в днях

In [10]:
days_employed_un = df_work['days_employed'].value_counts() # уникальные значения в столбце
days_employed_pass = df_work['days_employed'].isna().sum() # подсчёт пропусков

print(days_employed_un)
print(days_employed_pass)
days_employed_length = len(df_work['days_employed']) # длина столбца данных

share_of_pass = days_employed_pass/days_employed_length

print('Доля пропусков в столбце days_employed:{:.0%}'.format(share_of_pass))

-986.927316     1
-7026.359174    1
-4236.274243    1
-6620.396473    1
-1238.560080    1
               ..
-2849.351119    1
-5619.328204    1
-448.829898     1
-1687.038672    1
-582.538413     1
Name: days_employed, Length: 19351, dtype: int64
2174
Доля пропусков в столбце days_employed:10%


In [11]:
# дадим оценку отрицательным значениям в столбце days_employed
quantity_of_negative_indicator = df_work.loc[df_work['days_employed'] <= 0].count()# количество отрицательных показателей
print(quantity_of_negative_indicator)

share_of_quantity_of_negative_indicator = quantity_of_negative_indicator/days_employed_length # доля отрицательных показателей
print(share_of_quantity_of_negative_indicator)


children            15906
days_employed       15906
dob_years           15906
education           15906
education_id        15906
family_status       15906
family_status_id    15906
gender              15906
income_type         15906
debt                15906
total_income        15906
purpose             15906
dtype: int64
children            0.738955
days_employed       0.738955
dob_years           0.738955
education           0.738955
education_id        0.738955
family_status       0.738955
family_status_id    0.738955
gender              0.738955
income_type         0.738955
debt                0.738955
total_income        0.738955
purpose             0.738955
dtype: float64


**Вывод**: установлено наличие 2174 пропусков, что соответствует 10%. Доля отрицательных показателей в столбце days_employed:74% - для целей настоящего исследования отрицательные и невероятно завышенные показатели в столбце с указанием стажа не имеют значения, однако, необходимо обратиться к разрабочикам, дабы внести ясность.

In [12]:
# Заменим отрицательные значения на положительные
df_work['days_employed']=df_work['days_employed'].astype('str')
df_work['days_employed']=df_work['days_employed'].str.replace('-','')
df_work['days_employed']=df_work['days_employed'].astype('float')
df_work.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Так как между максимальными и минимальными значениями в столбце days_employed наблюдается значительный разброс,применим медианное значение.

In [13]:
# рассчитаем медианное значение и произведём замену в столбце days_employed
days_employed_median = df_work['days_employed'].median()
print(days_employed_median)
df_work['days_employed'] = df_work['days_employed'].fillna(value = days_employed_median)
df_work.head()

2194.220566878695


children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [14]:
# проверим себя
print(df_work['days_employed'].value_counts())
print(df_work['days_employed'].isna().sum())
print(len(df_work['days_employed']))

2194.220567    2175
986.927316        1
1893.222792       1
4236.274243       1
6620.396473       1
               ... 
2849.351119       1
5619.328204       1
448.829898        1
1687.038672       1
582.538413        1
Name: days_employed, Length: 19351, dtype: int64
0
21525


Пропуски устранены.

2.3.Получим информацию об уникальных значениях столбца dob_years — возраст клиента в годах

In [15]:
print(df_work['dob_years'].value_counts().sort_values())
print(df_work['dob_years'].isna().sum()) # вернёт количество пропусков;
print(len(df_work['dob_years'])) # покажет длину столбца;

75      1
74      6
73      8
19     14
72     33
20     51
71     58
70     65
69     85
68     99
0     101
21    111
67    167
22    183
66    183
65    194
23    254
24    264
64    265
63    269
62    352
61    355
25    357
60    377
26    408
55    443
59    444
51    448
53    459
57    460
58    461
46    475
54    479
47    480
52    484
56    487
27    493
45    497
28    503
49    508
32    510
43    513
50    514
37    537
48    538
30    540
29    545
44    547
36    555
31    560
39    573
33    581
42    597
38    598
34    603
41    607
40    609
35    617
Name: dob_years, dtype: int64
0
21525


в датафрейме в столбце с указанием возраста найден нулевой показатель, подсчитаем количество

In [16]:
quanity_of_zero = df_work.loc[df_work.loc[:,'dob_years'] == 0]['dob_years'].count()
print(quanity_of_zero)

101


In [17]:
print('Доля лиц с указанным возрастом - 0  в столбце dob_years:{:.2%}'.format(quanity_of_zero/len(df_work['dob_years'])))

Доля лиц с указанным возрастом - 0  в столбце dob_years:0.47%


Так как между максимальными и минимальными значениями в столбце dob_years наблюдается значительный разброс,применим медианное значение.

In [18]:
dob_years_median = int(df_work['dob_years'].median())
print(dob_years_median)

42


In [19]:
print(df_work.loc[df_work['dob_years'] == 0].count()) 
        

children            101
days_employed       101
dob_years           101
education           101
education_id        101
family_status       101
family_status_id    101
gender              101
income_type         101
debt                101
total_income         91
purpose             101
dtype: int64


In [20]:
df_work.loc[df_work['dob_years'] == 0,'dob_years'] = 42
print(df_work.loc[df_work['dob_years'] == 0].count()) 

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


Результат: нулевых показателей нет.

2.4.Получим информацию об уникальных значениях столбца education — уровень образования клиента

In [21]:
print(df_work['education'].value_counts()) # вернёт уникальные значения и их количество;
print(df_work['education'].isna().sum()) # вернёт количество пропусков;
print(len(df_work['education'])) # покажет длину столбца

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64
0
21525


Данные содержат различные способы упоминания одного и того же явления. Приведём к единому виду.

In [22]:
df_work['education']=df_work['education'].str.lower()
print (df_work['education'].value_counts())

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64


2.5.Получим информацию об уникальных значениях столбца education_id — идентификатор уровня образования

In [23]:
print(df_work['education_id'].value_counts().sort_values()) # вернёт уникальные значения и их количество;
print(df_work['education_id'].isna().sum()) # вернёт количество пропусков;

4        6
3      282
2      744
0     5260
1    15233
Name: education_id, dtype: int64
0


Все значения уникальные, пропусков нет.

2.6.Получим информацию об уникальных значениях столбца family_status — семейное положение

In [24]:
print(df_work['family_status'].value_counts()) # вернёт уникальные значения и их количество;
print(df_work['family_status'].isna().sum()) # вернёт количество пропусков;
print(len(df_work['family_status'])) # покажет длину столбца

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64
0
21525


In [25]:
# Доля каждой группы:
print(df.groupby(df_work['family_status']).size() / len(df_work['family_status']))

family_status
Не женат / не замужем    0.130685
в разводе                0.055517
вдовец / вдова           0.044599
гражданский брак         0.194053
женат / замужем          0.575145
dtype: float64


Все значения уникальные, пропусков нет. Самая значительная группа 'женат / замужем ' - 57.5%, далее, лица, состоящие в гражданском браке - 19,4%, Не женат / не замужем - 13%, в разводе - 5,5%, вдовец / вдова - 4,4%

2.7.Получим информацию об уникальных значениях столбца family_status_id — идентификатор семейного положения 

In [26]:
print(df_work['family_status_id'].value_counts().sort_values()) # вернёт уникальные значения и их количество;
print(df_work['family_status_id'].isna().sum()) # вернёт количество пропусков;
print(len(df_work['family_status_id'])) # покажет длину столбца

2      960
3     1195
4     2813
1     4177
0    12380
Name: family_status_id, dtype: int64
0
21525


Все значения уникальные, пропусков нет.

2.8.Получим информацию об уникальных значениях столбца gender — пол клиента

In [27]:
print(df_work['gender'].value_counts()) 
print(df_work['gender'].isna().sum())
print(len(df_work['gender'])) 

F      14236
M       7288
XNA        1
Name: gender, dtype: int64
0
21525


In [28]:
# так как XNA встретился 1 раз, приобщим к самой многочисленной группе F
df_work.loc[df_work['gender'] == 'XNA', 'gender'] = 'F'
print (df_work['gender'].value_counts())

F    14237
M     7288
Name: gender, dtype: int64


Аномалий нет

2.9.Получим информацию об уникальных значениях столбца income_type — тип занятости

In [29]:
print(df_work['income_type'].value_counts()) 
print(df_work['income_type'].isna().sum())
print(len(df_work['income_type'])) 

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64
0
21525


In [30]:
# так как наименование предприниматель,безработный встретились дважды
# студент, в декрете - однократно, приобщим эти записи к самой многочисленной группе - сотрудник
df_work.loc[df_work['income_type'] == 'предприниматель', 'income_type'] = 'сотрудник'
df_work.loc[df_work['income_type'] == 'студент', 'income_type'] = 'сотрудник'
df_work.loc[df_work['income_type'] == 'в декрете', 'income_type'] = 'сотрудник'
df_work.loc[df_work['income_type'] == 'безработный', 'income_type'] = 'сотрудник'
print (df_work['income_type'].value_counts())

сотрудник      11125
компаньон       5085
пенсионер       3856
госслужащий     1459
Name: income_type, dtype: int64


2.10.Получим информацию об уникальных значениях столбца debt — имел ли задолженность по возврату кредитов

In [31]:
print(df_work['debt'].value_counts()) 
print(df_work['debt'].isna().sum())
print(len(df_work['debt'])) 

0    19784
1     1741
Name: debt, dtype: int64
0
21525


Все значения уникальные, пропусков нет. Люди чаще возвращают кредиты.

2.11.Получим информацию об уникальных значениях столбца total_income — ежемесячный доход

In [32]:
print(df_work['total_income'].value_counts()) 
print(df_work['total_income'].isna().sum())
print(len(df_work['total_income'])) 

112874.418757    1
133912.272223    1
182036.676828    1
122421.963500    1
198271.837248    1
                ..
133299.194693    1
115080.782380    1
84896.781597     1
153838.839212    1
150014.128510    1
Name: total_income, Length: 19351, dtype: int64
2174
21525


Столбец total_income содержит такое же количество пропусков - 2174 как и days_employed, что соответствует 10% от общего количества наблюдений. Скорее всего, имеет место технический сбой (только в этих столбцах, к примеру, отражены данные типа float64). Можно предположить, что у человека нет стажа работы и дохода тоже нет. Хотя дожить на постсоветском пространстве до 63, 65 лет (строки 12, 29) со средним образованием и не иметь стажа и пенсии - маловероятная ситуация. Необходимо обратиться к разработчикам. В настоящий момент данные сведения не оказывают прямого воздействия на результат исследования. Произведём замену пропусков. 

In [33]:
# найдём предельные значения для столбца total_income, чтобы определиться с методом для замены пропусков.
print(df_work['total_income'].min())
print(df_work['total_income'].max())

20667.26379327158
2265604.028722744


Так как между максимальными и минимальными значениями наблюдается значитальный разброс, применим медианное значение.

In [34]:
# рассчитаем медианное значение и произведём замену в столбце total_income
total_income_median = df_work['total_income'].median()
print(int(total_income_median))
df_work['total_income'] = df_work['total_income'].fillna(value = total_income_median)
print(df_work['total_income'].isna().sum())

145017
0


Пропуски устранены. Произведём замену вещественого типа данных в столбце total_income на целочисленный методом astype().

In [35]:
df_work['total_income']=df_work['total_income'].astype('int')
df_work.dtypes

children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income          int64
purpose              object
dtype: object

Теперь значения целочисленные.

2.12.Получим информацию об уникальных значениях столбца purpose — цель получения кредита

In [36]:
print(df_work['purpose'].value_counts()) 
print(df_work['purpose'].isna().sum())
print(len(df_work['purpose'])) 

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Содержатся различные способы упоминания, представления одного и того же явления. Позже проведём категоризацию.

### Обработка дубликатов

Просмотрим весь датафрейм на предмет дубликатов методом duplicated(), который позволит получить строки с дубликатами. Основные причины возникновения дубликатов — повторные представления, различные способы упоминания, представления одного и того же явления\события, неправильное соединение данных из разных источников, ошибки пользователя при занесении информации. Отсюда и вытекает необходимость предобработки данных.

In [37]:
df_work[df_work.duplicated()]

children  days_employed  dob_years education  education_id  \
2849          0    2194.220567         41   среднее             1   
3290          0    2194.220567         58   среднее             1   
4182          1    2194.220567         34    высшее             0   
4851          0    2194.220567         60   среднее             1   
5557          0    2194.220567         58   среднее             1   
...         ...            ...        ...       ...           ...   
20702         0    2194.220567         64   среднее             1   
21032         0    2194.220567         60   среднее             1   
21132         0    2194.220567         47   среднее             1   
21281         1    2194.220567         30    высшее             0   
21415         0    2194.220567         54   среднее             1   

          family_status  family_status_id gender income_type  debt  \
2849    женат / замужем                 0      F   сотрудник     0   
3290   гражданский брак                 1      F   пенсионер     0   
4182   гражданский брак                 1      F   сотрудник     0   
4851   гражданский брак                 1      F   пенсионер     0   
5557   гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
20702   женат / замужем                 0      F   пенсионер     0   
21032   женат / замужем                 0      F   пенсионер     0   
21132   женат / замужем                 0      F   сотрудник     0   
21281   женат / замужем                 0      F   сотрудник     0   
21415   женат / замужем                 0      F   пенсионер     0   

       total_income                            purpose  
2849         145017            покупка жилья для семьи  
3290         145017                    сыграть свадьбу  
4182         145017                            свадьба  
4851         145017                            свадьба  
5557         145017                    сыграть свадьбу  
...             ...                                ...  
20702        145017         дополнительное образование  
21032        145017              заняться образованием  
21132        145017                       ремонт жилью  
21281        145017  покупка коммерческой недвижимости  
21415        145017                  операции с жильем  

[72 rows x 12 columns]

In [38]:
df.duplicated().sum()

72

In [39]:
print('Доля дубликатов в датафрейме:{:.2%}'.format(df.duplicated().sum()/len(df_work['days_employed'])))

Доля дубликатов в датафрейме:0.33%


**Вывод**: установлено наличие дубликатов, доля дубликатов в датафрейме:0.33%. Устраним их. 

In [40]:
df_work = df_work.drop_duplicates().reset_index(drop=True)
print(len(df_work))

21453


### Категоризация данных

In [41]:
def total_income_category(total_income):
    if 0 <= total_income <= 30000:
        return 'категория E'
    if 30001 <= total_income <= 50000:
        return 'категория D'
    if 50001 <= total_income <= 200000:
        return 'категория C'
    if 200001 <= total_income <= 1000000:
        return 'категория B'

    return 'категория A' 

df_work['total_income_category'] = df_work['total_income'].apply(total_income_category)
print(df_work.head(10))

   children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   среднее             1   
8         2    6929.865299         35    высшее             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M

In [42]:
print(total_income_category(32572)) # проверка

категория D


Замена произведена корректно.

In [43]:
# Доля каждой группы:
print(df_work.groupby(df_work['total_income_category']).size() / len(df_work['total_income_category']))

total_income_category
категория A    0.001165
категория B    0.234979
категория C    0.746516
категория D    0.016315
категория E    0.001025
dtype: float64


клиенты, чей доход находится в диапазоне от 0 до 30 000 - категорию 'E'- 0,1%; 
при доходе от 30 001 до 50 000 — категория 'D' - 1,6% ; 
50 001–200 000 — категория 'C' - 74,6% ; 
200 001–1 000 000 — категория 'B' - 23,4%; 
от 1 000 001 - категория 'A' -0,1%.

Для столбца **purpose** определим следующие категории: 'операции с автомобилем', 'операции с недвижимостью', 'проведение свадьбы', 'получение образования'. Для обощения наименований в категории, выполним отбор и группировку:

In [44]:
# Для категории  'операции с автомобилем' создадим переменную avto, 
#которая будет хранить варианты написания подобной причины из столбца purpose

avto =[]
queries = ['свадьба', 'на проведение свадьбы', 'сыграть свадьбу', 
        'операции с недвижимостью', 'покупка коммерческой недвижимости', 
        'покупка жилья для сдачи', 'операции с жильем', 'операции с коммерческой недвижимостью', 
        'покупка жилья', 'жилье', 'покупка жилья для семьи' , 
        'строительство собственной недвижимости', 'недвижимость', 
        'операции со своей недвижимостью', 'строительство жилой недвижимости' , 
        'покупка недвижимости', 'строительство недвижимости', 'покупка своего жилья', 
        'ремонт жилью', 'покупка жилой недвижимости', 'на покупку своего автомобиля', 
        'заняться высшим образованием', 'автомобиль' , 'сделка с подержанным автомобилем' , 
        'свой автомобиль' , 'на покупку подержанного автомобиля' , 'автомобили', 'на покупку автомобиля', 
        'дополнительное образование', 'приобретение автомобиля', 'сделка с автомобилем', 
        'высшее образование' , 'образование', 'получение дополнительного образования', 
        'получение образования', 'профильное образование', 'получение высшего образования', 
        'заняться образованием' ]
for query in queries:
    for word in query.split():
        stemmed_word = russian_stemmer.stem(word)
        if stemmed_word == 'автомобил':
            avto.append(query)
print(avto) 

['на покупку своего автомобиля', 'автомобиль', 'сделка с подержанным автомобилем', 'свой автомобиль', 'на покупку подержанного автомобиля', 'на покупку автомобиля', 'приобретение автомобиля', 'сделка с автомобилем']


In [45]:
# Для категории  'проведение свадьбы' создадим переменную wedding, 
#которая будет хранить варианты написания подобной причины из столбца purpose

wedding =[]
for query in queries:
    for word in query.split():
        stemmed_word = russian_stemmer.stem(word)
        if stemmed_word == 'свадьб':
            wedding.append(query)
print(wedding) 

['свадьба', 'на проведение свадьбы', 'сыграть свадьбу']


In [46]:
# Для категории  'получение образования' создадим переменную education, 
#которая будет хранить варианты написания подобной причины из столбца purpose

education =[]
for query in queries:
    for word in query.split():
        stemmed_word = russian_stemmer.stem(word)
        if stemmed_word == 'образован':
            education.append(query)
print(education)            

['заняться высшим образованием', 'дополнительное образование', 'высшее образование', 'образование', 'получение дополнительного образования', 'получение образования', 'профильное образование', 'получение высшего образования', 'заняться образованием']


In [47]:
# Для категории  'операции с недвижимостью' создадим переменную realty, 
#которая будет хранить варианты написания подобной причины из столбца purpose
realty = []
for query in queries:
    for word in query.split():
        stemmed_word = russian_stemmer.stem(word)
        if stemmed_word == 'недвижим':
            realty.append(query)
print(realty) 

['операции с недвижимостью', 'покупка коммерческой недвижимости', 'операции с коммерческой недвижимостью', 'строительство собственной недвижимости', 'недвижимость', 'операции со своей недвижимостью', 'строительство жилой недвижимости', 'покупка недвижимости', 'строительство недвижимости', 'покупка жилой недвижимости']


In [48]:
# для вариаций с упоминанием "жильё"
for query in queries:
    for word in query.split():
        stemmed_word = russian_stemmer.stem(word)
        if stemmed_word == 'жил':
            realty.append(query)
print(realty)  

['операции с недвижимостью', 'покупка коммерческой недвижимости', 'операции с коммерческой недвижимостью', 'строительство собственной недвижимости', 'недвижимость', 'операции со своей недвижимостью', 'строительство жилой недвижимости', 'покупка недвижимости', 'строительство недвижимости', 'покупка жилой недвижимости', 'покупка жилья для сдачи', 'операции с жильем', 'покупка жилья', 'жилье', 'покупка жилья для семьи', 'строительство жилой недвижимости', 'покупка своего жилья', 'ремонт жилью', 'покупка жилой недвижимости']


Спасибо, так намного проще, я применила ниже

In [49]:
# Функция purpose_category присвоит соответствующие категории, далее добавим столбец purpose_category в датафрейм
def purpose_category(purpose):
    if purpose in wedding:
        return 'проведение свадьбы'
    if purpose in education:
        return 'получение образования'
    if purpose in realty:
        return 'операции с недвижимостью'
    if purpose in avto:
        return 'операции с автомобилем'


df_work['purpose_category'] = df_work['purpose'].apply(purpose_category)
print(df_work.head(10))

   children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   среднее             1   
8         2    6929.865299         35    высшее             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M

In [50]:
# проверим корректность на примере
print(purpose_category('заняться высшим образованием'))


получение образования


In [51]:
def purpose_category_rev1(purpose):
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    elif 'жиль' in purpose or 'недвиж' in purpose:
        return 'операции с недвижимостью'
    elif 'авто' in purpose:
        return 'операции с автомобилем'
    elif 'образов' in purpose:
        return 'получение образования'
    else:
        return 'прочее' 

df_work['purpose_category_rev'] = df_work['purpose'].apply(purpose_category_rev1)
df_work['purpose_category_rev'].value_counts() 
print(df_work.head(10))

   children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   среднее             1   
8         2    6929.865299         35    высшее             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M

In [52]:
# Доля каждой группы:
print(df_work.groupby(df_work['purpose_category']).size() / len(df_work['purpose_category']))

purpose_category
операции с автомобилем      0.178437
операции с недвижимостью    0.503939
получение образования       0.187060
проведение свадьбы          0.108283
dtype: float64


**Вывод**: функции для присвоения категорий сработали корректно. 
Назначения кредитов распределились в следующей пропорции:
операции с недвижимостью    50%
получение образования       18,7%
операции с автомобилем      17,8%
проведение свадьбы          10%

### Содание дополнительных таблиц

Продолжим работу по оптимизации данных. Обратим внимание на группу столбцов family_status, family_status_id, education и education_id. 
Создадим новые датафреймы

**family_dict**, куда войдёт информация из family_status, family_status_id.

**educ_dict**, куда поместим сведения из education и education_id.

**purpose_dict**, куда поместим сведения из purpose и purpose_category.

**total_income_dict**, куда поместим сведения из total_income и total_income_category.

В рабочем датафрейме оставим family_status_id, education_id, purpose_category, total_income_category, что облегчит восприятие, а сформированные датафреймы могут служить источником справочной информации при необходимости.

In [53]:
df_work.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 15 columns):
children                 21453 non-null int64
days_employed            21453 non-null float64
dob_years                21453 non-null int64
education                21453 non-null object
education_id             21453 non-null int64
family_status            21453 non-null object
family_status_id         21453 non-null int64
gender                   21453 non-null object
income_type              21453 non-null object
debt                     21453 non-null int64
total_income             21453 non-null int64
purpose                  21453 non-null object
total_income_category    21453 non-null object
purpose_category         20975 non-null object
purpose_category_rev     21453 non-null object
dtypes: float64(1), int64(6), object(8)
memory usage: 2.5+ MB


In [54]:
family_dict = df_work[['family_status', 'family_status_id']]
family_dict = family_dict.drop_duplicates()
df_work = df_work.drop(columns=['family_status'])

purpose_dict = df_work [['purpose','purpose_category']]
purpose_dict = purpose_dict.drop_duplicates()
df_work = df_work.drop(columns=['purpose'])

total_income_dict = df_work [['total_income','total_income_category']]
total_income_dict = total_income_dict.drop_duplicates()
df_work = df_work.drop(columns=['total_income'])

educ_dict = df_work[['education_id', 'education']]
educ_dict = educ_dict.drop_duplicates()
df_work = df_work.drop(columns=['education'])
df_work = df_work.drop(columns=['purpose_category_rev'])

In [55]:
df_work.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 10 columns):
children                 21453 non-null int64
days_employed            21453 non-null float64
dob_years                21453 non-null int64
education_id             21453 non-null int64
family_status_id         21453 non-null int64
gender                   21453 non-null object
income_type              21453 non-null object
debt                     21453 non-null int64
total_income_category    21453 non-null object
purpose_category         20975 non-null object
dtypes: float64(1), int64(5), object(4)
memory usage: 1.6+ MB


In [56]:
# проверим не появились ли дубликаты
df_work.duplicated().sum()

331

In [57]:
print('Доля дубликатов в датафрейме:{:.2%}'.format(df_work.duplicated().sum()/len(df_work['days_employed'])))

Доля дубликатов в датафрейме:1.54%


In [58]:
# очистим данные от дубликатов
df_work = df_work.drop_duplicates().reset_index(drop=True)
print(len(df_work))

21122


## Ответы на вопросы, поставленные в начале исследования

**Есть ли зависимость между наличием детей и возвратом кредита в срок?**

In [59]:
df_children = df_work.groupby('children')['debt'].agg(['count', 'sum'])
df_children.columns = ['Общее число лиц в группе', 'Количество должников в группе']
df_children

Общее число лиц в группе  Количество должников в группе
children                                                         
0                            13831                           1061
1                             4798                            445
2                             2115                            202
3                              329                             27
4                               40                              4
5                                9                              0

In [60]:
print(df_work.loc[df_work.loc[:,'debt'] == 1]['debt'].count())

1739


In [61]:
df_children = df_work.groupby('children')['debt'].agg(['sum'])
df_children


sum
children      
0         1061
1          445
2          202
3           27
4            4
5            0

In [62]:
share_of_group_children = df_children/df_work.loc[df_work.loc[:,'debt'] == 1]['debt'].count()
share_of_group_children

sum
children          
0         0.610121
1         0.255894
2         0.116159
3         0.015526
4         0.002300
5         0.000000

**Вывод**: 

заёмщики, у которых нет детей чаще не возвращают заёмные средства - 61% случаев несвоевременных платежей;

заёмщики, у которых 1 ребёнок составляют долю 25,5%;

заёмщики, у которых 2 ребёнка составляют долю 11,6%;

заёмщики, у которых от 3 до 5 детей - самые безопасные - составляют долю 1,7%.


**Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [63]:
df_marrige = df_work.groupby('family_status_id')['debt'].agg(['count', 'sum'])
df_marrige.columns = ['Общее число лиц в группе', 'Количество должников в группе']
df_marrige

Общее число лиц в группе  Количество должников в группе
family_status_id                                                         
0                                    12078                            929
1                                     4123                            388
2                                      943                             63
3                                     1193                             85
4                                     2785                            274

In [64]:
df_marrige = df_work.groupby('family_status_id')['debt'].agg(['sum'])
share_of_group_marrige = df_marrige/df_work.loc[df_work.loc[:,'debt'] == 1]['debt'].count()
share_of_group_marrige

sum
family_status_id          
0                 0.534215
1                 0.223117
2                 0.036228
3                 0.048879
4                 0.157562

**Вывод**: 

клиенты, состоящие в браке, намного чаще допускают просрочки при платежах - 53% случаев.

Клиенты, состоящие в гражданском браке, занимают долю - 22% от общего количества несвоевременных платежей.

Вдовцы занимают долю - 3,6% от общего количества несвоевременных платежей.

Клиенты, находящиеся в разводе, занимают долю - 5% от общего количества несвоевременных платежей.

Клиенты, несостоящие в браке, занимают долю - 15,7% от общего количества несвоевременных платежей. 

**Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [65]:
df_income = df_work.groupby('total_income_category')['debt'].agg(['count', 'sum'])
df_income.columns = ['Общее число лиц в группе', 'Количество должников в группе']
df_income

Общее число лиц в группе  Количество должников в группе
total_income_category                                                         
категория A                                  25                              2
категория B                                5041                            356
категория C                               15684                           1358
категория D                                 350                             21
категория E                                  22                              2

In [66]:
df_income = df_work.groupby('total_income_category')['debt'].agg(['sum'])
share_of_group_income = df_income/df_work.loc[df_work.loc[:,'debt'] == 1]['debt'].count()
share_of_group_income

sum
total_income_category          
категория A            0.001150
категория B            0.204715
категория C            0.780909
категория D            0.012076
категория E            0.001150

**Вывод**: 
самая неблагонадёжная группа заёмщиков - клиенты с доходом от 50 001 до 200 000. На данную группу приходится 78% случаев несвоевременной оплаты.

20% случаев несвоевременной оплаты приходится на лиц с доходом от 200 001–1 000 000 — категория 'B'

1.2% случаев несвоевременной оплаты приходится на лиц с доходом от 30 000 до 50 000 - категория 'D'

категория 'A'(от 1 000 001) и категория 'E' (0 до 30 000) - самые благонадёжные

**Как разные цели кредита влияют на его возврат в срок?**

In [67]:
df_purpose = df_work.groupby('purpose_category')['debt'].agg(['count', 'sum'])

df_purpose.columns = ['Общее число лиц в группе', 'Количество должников в группе']
df_purpose


Общее число лиц в группе  \
purpose_category                                     
операции с автомобилем                        3800   
операции с недвижимостью                     10575   
получение образования                         3964   
проведение свадьбы                            2305   

                          Количество должников в группе  
purpose_category                                         
операции с автомобилем                              358  
операции с недвижимостью                            781  
получение образования                               370  
проведение свадьбы                                  186

In [68]:
df_purpose = df_work.groupby('purpose_category')['debt'].agg(['sum'])
share_of_group_purpose = df_purpose/df_work.loc[df_work.loc[:,'debt'] == 1]['debt'].count()
share_of_group_purpose

sum
purpose_category                  
операции с автомобилем    0.205865
операции с недвижимостью  0.449109
получение образования     0.212766
проведение свадьбы        0.106958

**Вывод**: 

наиболее рисковым направлением получения заёмных средств является - операции с недвижимостью - 45% случаев несвоевременных платежей;

на втором месте - получение образования - 21%;

операции с автомобилем - 20,5%;

проведение свадьбы - 11%;

Применение сводных таблиц являлось одним из заданий. Предложенный тобой метод мне нравится намного больше. Спасибо. Однако, на мой взгляд, расчёт процентов в последнем столбце даёт некорректные результаты: получается, что самая многочисленная во всех отношениях группа имеет меньшее количество % должников, чем самая малочисленная: 7,5% против 9,7%.  Я считаю, что правильнее соотносить Количество должников в группе с общим числом просрочек.

In [69]:
df_example = df_work.groupby('children')['debt'].agg(['count', 'sum', 'mean'])
df_example.columns = ['Кол-во пользователей', 'Кол-во должников', '% должников']
df_example

Кол-во пользователей  Кол-во должников  % должников
children                                                     
0                        13831              1061     0.076712
1                         4798               445     0.092747
2                         2115               202     0.095508
3                          329                27     0.082067
4                           40                 4     0.100000
5                            9                 0     0.000000

**Можно сделать общие следующие выводы:**

**В зависимости от количества детей:**

заёмщики, у которых нет детей (66% всех обратившихся за кредитом) чаще не возвращают заёмные средства - 61% случаев несвоевременных платежей;

заёмщики, у которых 1 ребёнок (22,6% от общего количества) составляют долю 25,5% от общей суммы несвоевременных платежей;

заёмщики, у которых 2 ребёнка (10% от общего количества)составляют долю 11,6% от общей суммы несвоевременных платежей;

заёмщики, у которых от 3 до 5 детей (1,4%)- самые безопасные - доля несвоевременных платежей - 1,7%.

**В зависимости от семейного положения:**

Клиенты, состоящие в браке (самая значительная группа 'женат / замужем ' - 57.5%), намного чаще допускают просрочки при платежах - 53% случаев, причём люди с доходом категории С (50 001–200 000) занимают долю - 40% от общего количества;

клиенты, находящиеся в гражданском браке (19,4%) - 22% от общего количества несвоевременных платежей;

клиенты, несостоящие в браке (группа "Не женат / не замужем" - 13%)- занимают долю - 15,7% от общего количества несвоевременных платежей. 

клиенты, относящиеся к группе в разводе (5,5% всех обратившихся за кредитом) - 3,6%;

клиенты, относящиеся к группе вдовец / вдова (4,4%) - 4,7%.

**В зависимости от дохода:**

Клиенты с доходом от 50 001 до 200 000 (доля группы 74,6% в общем объёме заёмщиков) - 76% случаев несвоевременной оплаты;

от 200 001 до 1 000 000 — категория 'B' (доля группы 23,4% ) - 20% случаев несвоевременной оплаты; 

при доходе от 30 001 до 50 000 — категория 'D' (доля группы 1,6%) - 1,1% ;

от 1 000 001 - категория 'A'(0,1%) - 0,1%;

от 0 до 30 000 -категория 'E' (0,1%) - 0,1%.

**Назначения кредитов распределились в следующей пропорции:**

операции с недвижимостью - 50% - доля случаев несвоевременной оплаты - 45%

получение образования - 18,7% - доля случаев несвоевременной оплаты - 21%

операции с автомобилем -17,8% - доля случаев несвоевременной оплаты - 21%

проведение свадьбы - 10% - доля случаев несвоевременной оплаты - 11%


Примечание: доля групп, указанная в скобках, рассчитана при работе с каждым соответствующим столбцом. Доля случаев несвоевременной оплаты рассчитана при помощи калькулятора по отношению к 1 741 случаю невсвоевременной оплаты ('debt').

## Общий вывод

**Все исследуемые показатели оказывают воздействие на возврат кредита в срок.**

**В зависимости от количества детей:**

заёмщики, у которых нет детей (66% всех обратившихся за кредитом) чаще не возвращают заёмные средства - 61% случаев несвоевременных платежей;

заёмщики, у которых от 3 до 5 детей (1,4%)- самые безопасные - доля несвоевременных платежей - 1,7%.

**В зависимости от семейного положения:**

клиенты, состоящие в браке (самая значительная группа 'женат / замужем ' - 57.5%), намного чаще допускают просрочки при платежах - 53% случаев;

клиенты, относящиеся к группе в разводе (5,5% всех обратившихся за кредитом) - 3,6% - реже, чем представители других групп допускают просрочку платежей;

**В зависимости от дохода:**

клиенты с доходом от 50 001 до 200 000 (доля группы 74,6% в общем объёме заёмщиков) - 76% случаев несвоевременной оплаты;

категория 'A' (от 1 000 001) и категория 'E' (от 0 до 30 000) - реже, чем представители других групп допускают просрочку платежей.

**Назначения кредитов распределились в следующей пропорции:**

операции с недвижимостью - 50% - доля случаев несвоевременной оплаты - 45% - самый высокий риск несвоевремнной оплаты;

проведение свадьбы - 10% - доля случаев несвоевременной оплаты - 11% - наименее рисковое направление.


**Резюме**
Для банка самыми безопасными по результатм анализа представленных данных явились лица, соответствующие следующим признакам: количество детей - от 3 до 5, находящиеся в разводе, с доходом от 0 до 30 000 или же от 1 000 001 в месяц, намеревающиеся провести свадьбу.

Для банка самыми рисковыми по результатм анализа представленных данных явились лица, соответствующие следующим признакам:  не имеющие детей, состоящие в браке, с доходом от 50 001 до 200 000 в месяц, намеревающиеся приобрести недвижимость.

Чаще всего за кредитом обращались лица, состоящие в браке и не имеющие детей, с доходом от 50 001 до 200 000 в месяц, наиболее рисковое назначение кредита - операции с недвижимостью.


При работе с датафреймом были выявлены пропуски в столбцах days_employed и total_income, составившие 10% общего объёма. Была произведена замена на медианный показатель, в связи со значительным разбросом данных в столбцах. Следует отметить, что именно в этих столбцах тип данных отличается от остальных - float64.    

Столбец days_employed — общий трудовой стаж в днях - имеет отрицательные значения - доля, которых - 74% ,а также максимальное значение - 401755.400475 дней!(вероятно количество указано в часах). Для целей настоящего исследования отрицательные и невероятно завышенные показатели в столбце с указанием стажа не имеют значения, однако, необходимо обратиться к разрабочикам, дабы внести ясность.

Столбец total_income содержит такое же количество пропусков - 2174 как и days_employed, что соответствует 10% от общего количества наблюдений. Скорее всего, имеет место технический сбой (только в этих столбцах, к примеру, отражены данные типа float64). Можно предположить, что у человека нет стажа работы и дохода тоже нет. Хотя дожить на постсоветском пространстве до 63, 65 лет (строки 12, 29) со средним образованием и не иметь стажа и пенсии - маловероятная ситуация. Необходимо обратиться к разработчикам. В настоящий момент данные сведения не оказывают прямого воздействия на результат исследования. Произведена замена пропусков медианным значением.

Показатели в children были откорректирован: доля отрицательного значения(-1) - 0,2%, полагаю, что имеет место "человеческий фактор" - просто заменены -1 на 1; доля значения 20 - 0,3% -"человеческий фактор" (может быть соскользнула рука при вводе, 2 и 0 расположены друг под другом). Заменены 20 на 2.

Столбец dob_years - возраст - содержал нулевой показатель, была произведена замена на медианный показатель в 101 случае, так как между максимальными и минимальными значениями в столбце dob_years наблюдается значительный разброс,применим медианное значение.

Столбец education — уровень образования клиента - содержит различные способы упоминания одного и того же уровня, поэтому были приведены к единому виду.

Столбцы total_income и purpose потребовали категоризации, так как содержали значительное количество записей, затрудняя проведение анализа. 

Можно предложить клиенту выбирать значения из списка, что позволит сократить время обработки и объём задействованной памяти.


**Спасибо за время, которое ты уделил моему проекту. Я узнала новое для себя. Мне было интересно.**

Воспользуюсь предложением задавать вопросы.

1. Метод sort_values() сортирует количество повторов в каждой группе, как отсортировать сами группы (первый столбец)?

2      960
3     1195
4     2813
1     4177
0    12380

2. Как перевести дроби в проценты
print(df['family_status'].value_counts())
print(df.groupby(df['family_status']).size() / len(df['family_status']))

получилось

Не женат / не замужем    0.130685
в разводе                0.055517
вдовец / вдова           0.044599
гражданский брак         0.194053
женат / замужем          0.57514

In [ ]:
# Ответ на первый вопрос - добавлять .sort_index()


df_work['dob_years'].value_counts().sort_index()

19     14
20     51
21    110
22    183
23    244
24    264
25    357
26    403
27    488
28    498
29    540
30    530
31    555
32    506
33    572
34    598
35    605
36    547
37    534
38    592
39    564
40    600
41    599
42    680
43    504
44    538
45    488
46    468
47    460
48    529
49    495
50    505
51    439
52    473
53    450
54    464
55    435
56    471
57    445
58    442
59    438
60    364
61    346
62    341
63    262
64    250
65    189
66    177
67    164
68     98
69     85
70     65
71     55
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [71]:
# ответ на второй вопрос - пришло в голову следующее: 
# твой код был: df.groupby(df['family_status']).size() / len(df['family_status'])
# он даёт Series 
# заворачиваем Series в Dataframe и к нему применяем .style.format() 

(df.groupby(df['family_status']).size() / len(df['family_status'])).to_frame().style.format('{:.2%}')